pubsub/tests/unit/subscriber_test.py

```diff
@@ -686,6 +692,30 @@ async def f(*args, **kwargs):
                 logging.WARNING,
                 'Ack failed for ack_id=ack_id_2') in caplog.record_tuples
 
+    @pytest.mark.asyncio
+    async def test_acker_graceful_shutdown(subscriber_client):
+
+        mock = MagicMock()
+
+        async def f(*args, **kwargs):
+            mock(*args, **kwargs)
+
+        subscriber_client.acknowledge = f
+        queue = asyncio.Queue()
+        acker_task = asyncio.ensure_future(
+            acker(
+                'fake_subscription',
+                queue,
+                subscriber_client,
+                0.1,
+                MagicMock()
+            )
+        )
+        await queue.put('ack_id_1')
+        await asyncio.wait_for(queue.join(), 1.)
+        mock.assert_called_with('fake_subscription', ack_ids=['ack_id_1'])
```

Comment:
Is this test conceptually different from basic positive case? It looks like what we want to test is that `queue.join()` will return when and only when the mock object is called.

Reply:
Yeah that's essentially the only difference. Could just throw that logic in the `test_acker_does_ack` test? Something like

```python
@pytest.mark.asyncio
async def test_acker_does_ack(subscriber_client):
    queue = asyncio.Queue()
    acker_task = asyncio.ensure_future(
        acker(
            'fake_subscription',
            queue,
            subscriber_client,
            0.0,
            MagicMock()
        )
    )
    await queue.put('ack_id')
    await queue.join()
    subscriber_client.acknowledge.assert_called_once_with(
        'fake_subscription', ack_ids=['ack_id'])
    acker_task.cancel()
```

Reply:
Ok done 👍